### ISSUE TITLE
No DEFAULT PRIVILEGES for schemas

### ISSUE THREAD

**Issue by michalk-k** [dbeaver/dbeaver#8111]
Postgres offers setting default privileges for schemas.
DBeaver doesn't show those settings (pgAdmin III does)

**Comment by uslss** [comment]
thanks for suggestion

**Comment by LonwoLonwo** [comment]
Hello @michalk-k 

In the latest DBeaver version,  we have such functionality:

![2024-03-12 18_03_33-DBeaver Ultimate 24 0 0 - test_role](https://github.com/dbeaver/dbeaver/assets/45152336/6314ecb8-f9bf-43b7-bbe7-be813e34b931)

![2024-03-12 18_03_16-DBeaver Ultimate 24 0 0 - test_role](https://github.com/dbeaver/dbeaver/assets/45152336/c00d6708-8328-48e1-a12e-f27ee9fcd729)

Is this what you are looking for?

**Comment by michal-bartak** [comment]
thank you for heads-up.

From the first glance of what I see, there is a lack of a role which defaults are created for.

Postgres applies default privileges only in the context of a role that issues the creation of a new object. Only defaults defined for this role are applied.

for example:

```ALTER DEFAULT PRIVILEGES FOR ROLE admin REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC;```

in the example above, the revoke will be issued only if a function is being created by a user named 'admin'

Syntax without the `FOR ROLE` part substitutes creation for the current user (current for the default creation, not for applying default to newly created object).

So, if I issue the statement below while being logged with user 'michal'

```ALTER DEFAULT PRIVILEGES REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC;```

the result will be equal to issuing:

```ALTER DEFAULT PRIVILEGES FOR ROLE michal REVOKE EXECUTE ON FUNCTIONS FROM PUBLIC;```


tldr missing FOR ROLE in sqls generated by DBeaver, as well as missing role selector in default privileges GUI makes it effectively useless. 

As a side note, right now I'm not sure a PUBLIC metarole might be used for `FOR ROLE` part. I will be back with the answer a while later:
Edit: No, the default privileges cannot be set for public meta-role. The query below fails with _role `public` does not exist_ error:
```ALTER DEFAULT PRIVILEGES FOR ROLE public IN SCHEMA myschema GRANT EXECUTE ON FUNCTIONS TO anotherrole```

**Comment by michal-bartak** [comment]
I just run DBeaver to look at that...
Where should I look, to verify the existence of default privileges? I mean this 'ALTER DEFAULT PRIVILEGES....` expression, or GUI for maintaining default privs.

I can see this "default privileges" string under Grant ALL/Revoke All buttons... but it does nothing.

Just for reference, the Postgresql stores default privileges in the `pg_catalog.pg_default_acl` table. Here is a query to retrieve them:

```
SELECT
    defaclrole::regrole::TEXT,
    defaclnamespace::regnamespace::TEXT, 
    defaclobjtype,
    CASE grantee WHEN 0 THEN 'public' ELSE grantee::regrole::TEXT END,
    privilege_type
FROM pg_catalog.pg_default_acl AS acl
JOIN aclexplode(acl.defaclacl) AS x ON TRUE
```

**Comment by LonwoLonwo** [comment]
And here we are again with this ticket: https://github.com/dbeaver/dbeaver/issues/11061

We do not have PUBLIC role support in PostgreSQL at the moment.

> Where should I look, to verify the existence of default privileges?

You can check the permission of an object in the Permissions tab. 
You can check the DDL for the role. 
You can click on the container folder in the Permission Editor of the Role tab.

**Comment by michal-bartak** [comment]
I've checked it again:

I issued the following statement against my testing db:

```sql
ALTER DEFAULT PRIVILEGES FOR ROLE sazky IN SCHEMA clients GRANT SELECT ON TABLES TO app_client;
ALTER DEFAULT PRIVILEGES FOR ROLE sazky IN SCHEMA clients GRANT SELECT ON TABLES TO public;
```

In Source pane I can see:
```sql
ALTER DEFAULT PRIVILEGES IN SCHEMA clients GRANT SELECT ON TABLES TO public;
ALTER DEFAULT PRIVILEGES IN SCHEMA clients GRANT SELECT ON TABLES TO app_client;
```
see image:
![image](https://github.com/dbeaver/dbeaver/assets/7868445/a3e6f01b-2f86-41d6-8879-4b1340324b21)

There is `FOR ROLE sazky` missing.

This resulting query is incomplete since it doesn't contain a role for which the default is created. 
My original statements mean that the default grants will be applied only if new tables are created within a session with the current user identified by "sazky" role. 
The query shown in the Source pane creates defaults (not applying grants to tables during their creation) for a role that is currently logged in, which is completely different from the statements I issued. aka bug: https://github.com/dbeaver/dbeaver/issues/23419

Postgresql doesn't provide the option to create a default for everyone. The default might be created for the current or specific user. Example

```ALTER DEFAULT PRIVILEGES FOR ROLE michal IN SCHEMA clients GRANT SELECT ON TABLES TO app_client;```
is equal to connect to the database as `michal` and then issuing 
```ALTER DEFAULT PRIVILEGES IN SCHEMA clients GRANT SELECT ON TABLES TO app_client;```
That's all.

In addition to that, the permissions pane for a schema doesn't show any defaults at all (in a place described as Default Privileges)
![image](https://github.com/dbeaver/dbeaver/assets/7868445/814db110-736a-4d0b-a667-42a52dfb5172)


There are still missing default privileges on schemas (default grants applied to schemas during creation). Those are represented on database level while there is no the Source pane for database.

BTW Permissions without a public role supported have to be considered flawful/unreliable thus useless. it'sg good that the public role may be found in generated SQL queries
